backup/crowbar-backup

```diff
@@ -210,7 +213,50 @@ function restore()
 
   # install cloud product (expects that the repos are present)
   zypper -n install -l -t pattern cloud_admin
-  restore_files
+
+  if ! rpm -q patterns-cloud-admin &> /dev/null; then
+    echo "patterns-cloud-admin package is not installed; please install with"
+    echo "\"zypper in -t pattern cloud_admin\" or \"zypper in patterns-cloud-admin\"."
+    exit 82
+  fi
+
+  restore_files var/lib/crowbar/backup
+
+  if [ -f /var/lib/crowbar/backup/barclamp-list ]; then
+    missing_barclamps=
+    for barclamp in $(sort -u /var/lib/crowbar/backup/barclamp-list); do
+      rpm -q "crowbar-barclamp-$barclamp" &> /dev/null && continue
+      test -d "/opt/dell/barclamps/$barclamp" && continue
+
+      # Try to convert _ to - in the barclamp name
+      barclamp_dash=$(echo $barclamp | sed "s/_/-/g")
+      rpm -q "crowbar-barclamp-${barclamp_dash}" &> /dev/null && continue
+      test -d "/opt/dell/barclamps/${barclamp_dash}" && continue
+
+      # FIXME: in the future, it might be nicer to ask the user what to do
+      echo "Trying to install barclamp $barclamp that is required by the backup."
+      zypper -n install crowbar-barclamp-$barclamp crowbar-barclamp-${barclamp_dash}
+
+      rpm -q "crowbar-barclamp-$barclamp" &> /dev/null && continue
+      rpm -q "crowbar-barclamp-${barclamp_dash}" &> /dev/null && continue
+
+      missing_barclamps="$missing_barclamps $barclamp"
+    done
+
+    if [ -n "$missing_barclamps" ]; then
+      echo "The following barclamps were installed on the system where the backup was"
+      echo "created and are not available on this system. Please install them."
+      echo "   $missing_barclamps"
+      exit 83
+    fi
+  fi
+
+  restore_files etc/crowbar
+  restore_files $ETCFILES
+
+  # set the correct hostname
+  RESTORE_HOSTNAME=$(cat /etc/HOSTNAME)
```

Comment:
The purpose of cat is to concatenate (or "catenate") files. If it's only one file, concatenating it with nothing at all is a waste of time, and costs you a process. See http://www.smallo.ruhr.de/award.html#cat

In general the times were we optimized runtimes of shell scripts are long over.. hence the smiley.